### ISSUE TITLE
Huge jump in tsc memory usage from 5.5.4 -> 5.6.2

### ISSUE THREAD

**Issue by Scalahansolo** [microsoft/TypeScript#60166]
### 🔎 Search Terms

"transformTime time", "compile time"

### 🕗 Version & Regression Information

- This changed between versions 5.5.4 and 5.6.2

### 💻 Code

`tsc` on 5.6.2 is taking up significantly more memory for unknown reasons. No other dep changes other than typescript upgrade. I've looked into various articles / guides to try and track down how to further investigate this, but thus far I have not found a way to really track what is newly taking up so much more memory. 

### 🙁 Actual behavior

After upgrading to 5.6, the memory usage of my `tsc` is more than double with no other changes to my `.tsconfig.json` file or the underlying code

![CleanShot 2024-10-07 at 22 35 37@2x](https://github.com/user-attachments/assets/e92dcbfd-29b6-4e7e-ac78-3b1807a03264)

### 🙂 Expected behavior

Before upgrade to 5.6, when running `tsc` on my project I get the following diagnostics.

![CleanShot 2024-10-07 at 22 35 01@2x](https://github.com/user-attachments/assets/ffd01e08-b5e7-439f-86aa-0af2b5424f77)

**Comment by RyanCavanaugh** [comment]
Very difficult to speculate on what this might be. https://github.com/microsoft/TypeScript/wiki/Performance has information on how to run profiling, or you can share a repro with us

**Comment by RyanCavanaugh** [comment]
The best tool for figuring out what's taking up more memory would be a heap snapshot, but a heap snapshot would also contain all of your code, so you'd have to diagnose it yourself.

The two runs don't quite look apples-to-apples - are you sure you're doing a clean build in both cases? I'm not sure how the first one reports 13s of transform time on a 12s execution (????)

**Comment by Scalahansolo** [comment]
Just to double check, I re ran these `tsc` runs just now. Only different in these two runs in the typescript version.

Typescript 5.5.4
![CleanShot 2024-10-08 at 16 32 42@2x](https://github.com/user-attachments/assets/aad977ed-4e6a-4d1e-aeef-81be897ce261)

Typescript 5.6.2
![CleanShot 2024-10-08 at 16 33 51@2x](https://github.com/user-attachments/assets/1e1ce6b4-2a5b-473c-af8c-d9ab6b5d9141)

I'll go down the route and see I cant get some heap snapshots and see if anything looks suspicious.

**Comment by jakebailey** [comment]
One thing you may consider is bisecting this with https://www.npmjs.com/package/every-ts, if the memory jump is clear-cut and not spread out over the entire dev cycle.

**Comment by Scalahansolo** [comment]
Maybe I'm misunderstanding, but in my case there isn't anything to bisect? Quite literally the only change from the two tsc runs above is changing the typescript version.

**Comment by jakebailey** [comment]
every-ts builds TS from source, so it bisects the repo and all of the changes we made between two versions.

**Comment by Scalahansolo** [comment]
Okay, so while I couldn't get `every-ts` to work. I just brute forced it and did a manual bisect. Here are some interesting numbers and the exact versions where there was a sizable memory jump. For the sake being forward looking, I also tested the latest 5.7 version, which actually saw a big memory drop, so it seems like maybe someone on the team found something already?

| Version            | Mem Used |
| ------------------ | -------- |
| 5.7.0-dev.20241008 | 2719119K |
| 5.6.3              | 5550198K |
| 5.6.0-dev.20240817 | 5546372K |
| 5.6.0-dev.20240629 | 3625724K |
| 5.5.4              | 2047049K |

**Comment by jakebailey** [comment]
I'd be interested in what's in your `.data/TypeScript` dir (if you `ls` it), but it is moderately off-topic. If you delete `.data` and to do an `every-ts` fetch, it may resolve itself.

**Comment by Scalahansolo** [comment]
After deleting the `.data` dir and just running `every-ts` I get the follow.

![CleanShot 2024-10-08 at 20 38 12@2x](https://github.com/user-attachments/assets/b403ae3c-1954-43bc-91cb-0159b44cd920)

**Comment by jakebailey** [comment]
Hm, I can only think that your version of `git` is out of date, or that you have a corporate proxy which is screwing with GitHub pulls. I can only suggest updating `git` or deleting and retrying. I've seen that error before, but not in a while.

**Comment by Scalahansolo** [comment]
I can maybe look into that. But is that necessary given the information I provide above with manually bisecting the releases?

**Comment by jakebailey** [comment]
The table seems to only narrow it down to a month, which is half of our three month dev cycle, so unfortunately doesn't say much.

If you'd like to go farther, you could manually build from source, then run `npm ci && npx hereby tsc && built/local/tsc.js -p path/to/tsconfig.json` while bisecting.

Otherwise if you think it's fixed in 5.7, then that's fine, in which case this could be closed.

**Comment by Scalahansolo** [comment]
I think maybe my table wasn't clear. 06/28 -> 06/29 is where the first major jump in memory usage happened, than the next big jump being on 08/16 -> 08/17

**Comment by jakebailey** [comment]
Ah, I see.

- `5.6.0-dev.20240628` to `5.6.0-dev.20240629` is 1b867c52b2a54cff51125bf915b5c8790334a376...cb98634f4915bbf7468d33a380a54a372fb94e26
- `5.6.0-dev.20240816` to `5.6.0-dev.20240817` is 856e472104bc8f000d936642b2f46ece260ffb96...d06fb82d7c79cf26893b50d7787af0fe2ffbe845
- `5.7.0-dev.20241007` to `5.7.0-dev.20241008` is 44331b9fa32999d4aade7433c10cd3ec8596957e...cd6c0a0b6b8659f49344951219c74a516fdec9a2

Nothing here jumps out at me.

**Comment by sheetalkamat** [comment]
May be its "noEmit" now reports declaration errors so calculates those errors. which didnt happen before and hence the extra time and memory spent in that?

**Comment by Scalahansolo** [comment]
To be super clear... I didnt test `5.7.0-dev.20241007` to `5.7.0-dev.20241008` to see when the memory dropped again. I can try to track down when the memory dropped. I was only hunting for the increases in memeory.

**Comment by RyanCavanaugh** [comment]
I don't think we have enough information to fix the bug. Even knowing a specific commit isn't really going to help since we don't have a way to analyze the root cuase.

**Comment by nickythorne** [comment]
Also noticing an increased memory usage when building my application (using nextjs). Originally [commented on an open issue nextjs](https://github.com/vercel/next.js/issues/76704#issuecomment-2707416535). But after downgrading typescript to `5.5.4` (from `5.8.2`) the out of memory error was resolved. 

Is there any way I can provide tracing/debug info to help with the analysis?

**Comment by jakebailey** [comment]
https://www.npmjs.com/package/pprof-it can be used to take a memory allocation profile, but looking t the above, those two memory usage counts are not very far apart and could be attributed to basically anything. Finding a specific place between those two releases that changed things would be more helpful.

But, everyone's issues are different, it seems pretty unlikely that your issue is the same as this one unless you're using the same exact libraries and see the same bump at the same versions. 5.5 to 5.8 is far too large of a range to debug.

**Comment by AlexeiDarmin-SA** [comment]
Hi all, I was able to recreate a simple reproduction of the issue and shared it here https://github.com/microsoft/TypeScript/issues/62660

At the core, it seems to stem from a difference in how the two versions of typescript handle certain kinds of memory allocation issues with large types that sometimes cause `The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.`
